### ISSUE TITLE
[alerting] Add slack notification capability.

### ISSUE THREAD

**Issue by manugarg** [cloudprober/cloudprober#466]
Cloudprober alerting currently support following notification targets:
* PagerDuty
* Email
* Command execution

We should add support for slack notifications as well.

**Comment by robpickerill** [comment]
I'll take this one, if you haven't already started it, Manu - thanks!

**Comment by manugarg** [comment]
Hi @robpickerill, I've not started yet, but I was looking at various ways to implement it. There is a slack Go package but the functionality we need is super small. We just to need to post a message to slack, similar to the pagerduty implementation but much smaller as it will be a generic message. So I am not sure if we should use the slack go-package.

Another thing I was thinking of changing, before publicizing the alerting functionality, was to group notifier specific configs in the NotifyConfig message:
https://github.com/cloudprober/cloudprober/blob/master/probes/alerting/proto/config.proto

Something like:
```
PagerDuty {
  ..
  ..
}

Slack {
  ..
}

NotifyConfig {
   .. 
  Email           email = 11
  PagerDuty  pager_duty = 12
  Slack           slack = 13
}
```

What do you think?

>> I'll take this one

Sounds good to me. Thanks Rob!

**Comment by robpickerill** [comment]
Sounds great, thanks Manu! 

I was considering this API: https://api.slack.com/methods/chat.postMessage, or this API: https://api.slack.com/messaging/webhooks. I agree, a simple http post is all that's needed here, so I was going to implement that with the go http client. Both appear to be viable, the first requires the user create a slack app and the token be provided to cloudprober, the latter requires the webhook app to be installed and configured, and the webhook provided to cloudprober.

Note: [https://api.slack.com/messaging/sending > ](https://api.slack.com/messaging/sending#sending_methods) details the methods we can send messages to slack, do you have a preference?

> group notifier specific configs
I like that idea, sounds great, would you like to me to take a look at that, as a pre-requisite to slack functionality?

**Comment by manugarg** [comment]
> Sounds great, thanks Manu!
> 
> I was considering this API: https://api.slack.com/methods/chat.postMessage, or this API: https://api.slack.com/messaging/webhooks. I agree, a simple http post is all that's needed here, so I was going to implement that with the go http client. Both appear to be viable, the first requires the user to create a Slack app and the token be provided to cloudprober, the latter requires the webhook app to be installed and configured, and the webhook provided to cloudprober.

I think app has be created regardless. Webhook is probably a better option for this -- that's what I've used with Grafana and PagerDuty before, e.g..
> 
> Note: [https://api.slack.com/messaging/sending](https://api.slack.com/messaging/sending#sending_methods) details the methods we can send messages to slack, do you have a preference?
> 
> > group notifier specific configs
> 
> I like that idea, sounds great, would you like for me to have a look at that, as a prerequisite to slack functionality?
For slack config we can go with this option, and change others independently (I'll follow up on that)?

**Comment by robpickerill** [comment]
Sounds great. I agree, the webhook is the easiest in my experience.

**Comment by manugarg** [comment]
@robpickerill implemented this in #475. Thanks Rob!
